File: api/example_usage.c
/**
 * @file example_usage.c
 * @brief Example usage of the Lumitec Poco CAN API
 * 
 * This example demonstrates how to use the Lumitec Poco API to create
 * and parse CAN messages for controlling Lumitec lighting systems.
 */

#include "lumitec_poco_api.h"
#include <stdio.h>
#include <string.h>

// Example function to print a CAN frame in hex format
void print_can_frame(const lumitec_poco_can_frame_t* frame)
{
    printf("CAN ID: 0x%08X, DLC: %d, Data: ", frame->can_id, frame->data_length);
    for (int i = 0; i < frame->data_length; i++) {
        printf("%02X ", frame->data[i]);
    }
    printf("\n");
}

int main(void)
{
    printf("Lumitec Poco CAN API Example\n");
    printf("API Version: %s\n\n", lumitec_poco_get_version());
    
    // Example 1: Create a simple action message
    printf("=== Example 1: Simple Action Message ===\n");
    lumitec_poco_can_frame_t frame;
    
    if (lumitec_poco_create_simple_action(&frame, 0x0E, 0x10, POCO_ACTION_ON, 1)) {
        printf("Created simple action message (Turn On, Switch 1):\n");
        print_can_frame(&frame);
        
        // Parse the message back
        lumitec_poco_simple_action_t action;
        if (lumitec_poco_parse_simple_action(&frame, &action)) {
            printf("Parsed: Action=%s, Switch=%d\n",
                   lumitec_poco_action_to_string(action.action_id),
                   action.switch_id);
        }
    }
    printf("\n");
    
    // Example 2: Create a custom HSB message
    printf("=== Example 2: Custom HSB Message ===\n");
    if (lumitec_poco_create_custom_hsb(&frame, 0x0E, 0x10, POCO_ACTION_T2HSB, 1, 128, 255, 200)) {
        printf("Created custom HSB message (Hue=128, Sat=255, Bright=200):\n");
        print_can_frame(&frame);
        
        // Parse the message back
        lumitec_poco_custom_hsb_t hsb;
        if (lumitec_poco_parse_custom_hsb(&frame, &hsb)) {
            printf("Parsed: Action=%s, Switch=%d, H=%d, S=%d, B=%d\n",
                   lumitec_poco_action_to_string(hsb.action_id),
                   hsb.switch_id, hsb.hue, hsb.saturation, hsb.brightness);
        }
    }
    printf("\n");
    
    // Example 3: Create a state information message
    printf("=== Example 3: State Information Message ===\n");
    if (lumitec_poco_create_state_info(&frame, 0x10, 2, POCO_SWITCH_STATE_PRESSED, POCO_SWITCH_TYPE_MOMENTARY)) {
        printf("Created state info message (Switch 2 pressed, momentary):\n");
        print_can_frame(&frame);
        
        // Parse the message back
        lumitec_poco_state_info_t state_info;
        if (lumitec_poco_parse_state_info(&frame, &state_info)) {
            printf("Parsed: Switch=%d, State=%s, Type=%s\n",
                   state_info.switch_id,
                   lumitec_poco_state_to_string(state_info.switch_state),
                   lumitec_poco_type_to_string(state_info.switch_type));
        }
    }
    printf("\n");
    
    // Example 4: Create a start pattern message
    printf("=== Example 4: Start Pattern Message ===\n");
    if (lumitec_poco_create_start_pattern(&frame, 0x0E, 0x10, 1, 5)) {
        printf("Created start pattern message (Switch 1, Pattern 5):\n");
        print_can_frame(&frame);
        
        // Parse the message back
        lumitec_poco_start_pattern_t pattern;
        if (lumitec_poco_parse_start_pattern(&frame, &pattern)) {
            printf("Parsed: Switch=%d, Pattern=%d\n",
                   pattern.switch_id, pattern.pattern_id);
        }
    }
    printf("\n");
    
    // Example 5: Validate frame
    printf("=== Example 5: Frame Validation ===\n");
    printf("Frame is valid Lumitec Poco message: %s\n",
           lumitec_poco_is_valid_frame(&frame) ? "Yes" : "No");
    
    uint8_t prop_id;
    if (lumitec_poco_get_proprietary_id(&frame, &prop_id)) {
        printf("Proprietary ID: %d\n", prop_id);
    }
    
    return 0;
}

File: api/python/example_usage.py
#!/usr/bin/env python3
"""
Example usage of the Lumitec Poco CAN API Python library

This example demonstrates how to use the Python Lumitec Poco API to create
and parse CAN messages for controlling Lumitec lighting systems.
"""

from lumitec_poco import (
    LumitecPocoAPI, PocoCANFrame, PocoActionID, PocoSwitchState, PocoSwitchType,
    create_simple_action, create_custom_hsb, create_state_info, create_start_pattern,
    is_valid_poco_frame, parse_poco_frame
)


def print_can_frame(frame: PocoCANFrame) -> None:
    """Print a CAN frame in hex format"""
    data_hex = frame.data.hex().upper()
    data_hex_spaced = ' '.join(data_hex[i:i+2] for i in range(0, len(data_hex), 2))
    print(f"CAN ID: 0x{frame.can_id:08X}, DLC: {frame.data_length}, Data: {data_hex_spaced}")


def main():
    print("Lumitec Poco CAN API Python Example")
    print(f"API Version: {LumitecPocoAPI.get_version()}\n")
    
    # Example 1: Create a simple action message
    print("=== Example 1: Simple Action Message ===")
    frame = create_simple_action(0x0E, 0x10, PocoActionID.ON, 1)
    print("Created simple action message (Turn On, Switch 1):")
    print_can_frame(frame)
    
    # Parse the message back
    action = LumitecPocoAPI.parse_simple_action(frame)
    if action:
        print(f"Parsed: Action={LumitecPocoAPI.action_to_string(PocoActionID(action.action_id))}, Switch={action.switch_id}")
    print()
    
    # Example 2: Create a custom HSB message
    print("=== Example 2: Custom HSB Message ===")
    frame = create_custom_hsb(0x0E, 0x10, PocoActionID.T2HSB, 1, 128, 255, 200)
    print("Created custom HSB message (Hue=128, Sat=255, Bright=200):")
    print_can_frame(frame)
    
    # Parse the message back
    hsb = LumitecPocoAPI.parse_custom_hsb(frame)
    if hsb:
        print(f"Parsed: Action={LumitecPocoAPI.action_to_string(PocoActionID(hsb.action_id))}, "
              f"Switch={hsb.switch_id}, H={hsb.hue}, S={hsb.saturation}, B={hsb.brightness}")
    print()
    
    # Example 3: Create a state information message
    print("=== Example 3: State Information Message ===")
    frame = create_state_info(0x10, 2, PocoSwitchState.PRESSED, PocoSwitchType.MOMENTARY)
    print("Created state info message (Switch 2 pressed, momentary):")
    print_can_frame(frame)
    
    # Parse the message back
    state_info = LumitecPocoAPI.parse_state_info(frame)
    if state_info:
        print(f"Parsed: Switch={state_info.switch_id}, "
              f"State={LumitecPocoAPI.state_to_string(PocoSwitchState(state_info.switch_state))}, "
              f"Type={LumitecPocoAPI.type_to_string(PocoSwitchType(state_info.switch_type))}")
    print()
    
    # Example 4: Create a start pattern message
    print("=== Example 4: Start Pattern Message ===")
    frame = create_start_pattern(0x0E, 0x10, 1, 5)
    print("Created start pattern message (Switch 1, Pattern 5):")
    print_can_frame(frame)
    
    # Parse the message back
    pattern = LumitecPocoAPI.parse_start_pattern(frame)
    if pattern:
        print(f"Parsed: Switch={pattern.switch_id}, Pattern={pattern.pattern_id}")
    print()
    
    # Example 5: Validate frame and auto-parse
    print("=== Example 5: Frame Validation and Auto-parsing ===")
    print(f"Frame is valid Lumitec Poco message: {is_valid_poco_frame(frame)}")
    
    prop_id = LumitecPocoAPI.get_proprietary_id(frame)
    if prop_id:
        print(f"Proprietary ID: {prop_id}")
    
    # Auto-parse any Poco frame
    parsed_msg = parse_poco_frame(frame)
    if parsed_msg:
        print(f"Auto-parsed message type: {type(parsed_msg).__name__}")
        print(f"Message content: {parsed_msg}")
    print()
    
    # Example 6: Working with different action types
    print("=== Example 6: Different Action Types ===")
    actions = [
        (PocoActionID.OFF, "Turn Off"),
        (PocoActionID.DIM_UP, "Dim Up"),
        (PocoActionID.RED, "Red Color"),
        (PocoActionID.PATTERN_START, "Start Pattern")
    ]
    
    for action_id, description in actions:
        frame = create_simple_action(0x0E, 0x10, action_id, 1)
        print(f"{description}: {frame.data.hex().upper()}")
    print()
    
    # Example 7: HSB Color Examples
    print("=== Example 7: HSB Color Examples ===")
    colors = [
        (0, 255, 255, "Pure Red"),
        (85, 255, 255, "Pure Green"),
        (170, 255, 255, "Pure Blue"),
        (0, 0, 255, "White"),
        (42, 255, 200, "Orange")
    ]
    
    for hue, sat, bright, description in colors:
        frame = create_custom_hsb(0x0E, 0x10, PocoActionID.T2HSB, 1, hue, sat, bright)
        print(f"{description} (H={hue}, S={sat}, B={bright}): {frame.data.hex().upper()}")


if __name__ == "__main__":
    main()
